src/humanloop/otel/helpers.py

```diff
@@ -0,0 +1,197 @@
+        key: Key prefix to write to the span attributes. The path to the values does not
+            need to exist in the span attributes.
+    """
+    to_write: dict[str, Any] = {}
+    _linear_object(to_write, value)
+    for k, v in to_write.items():
+        # OTT
+        if v is not None:
+            span._attributes[f"{key}.{k}" if key != "" else k] = v
+    # with _cache_lock:
+    #     _cache[(span.context.span_id, key)] = value
+
+
+def read_from_opentelemetry_span(span: ReadableSpan, key: str = ""):
+    """Read a value from the OpenTelemetry span attributes.
+
+    OpenTelemetry liniarises dictionaries and lists, storing only primitive values
+    in the span attributes. This function reconstructs the original structure from
+    a key prefix.
+
+    Arguments:
+        span: OpenTelemetry span to read values from
+        key: Key prefix to read from the span attributes
+
+    Returns:
+        Python object stored in the span attributes under the key prefix.
+
+    Examples:
+    `span.attributes` contains the following attributes:
+    ```python
+        foo.x.y = 7
+        foo.x.z.a = 'hello'
+        foo.x.z.b = 'world'
+        baz.0 = 42
+        baz.1 = 43
+    ```
+
+    `read_from_opentelemetry_span(span, key='foo')` returns:
+    ```python
+    {
+        'x': {
+            'y': 7,
+            'z': {
+                'a': 'hello',
+                'b': 'world'
+            }
+        }
+    }
+    ```
+
+    `read_from_opentelemetry_span(span, key='foo.x')` returns:
+    ```python
+    {
+        'y': 7,
+        'z': {
+            'a': 'hello',
+            'b': 'world'
+        }
+    }
+    ```
+
+    `read_from_opentelemetry_span(span, key='baz')` returns:
+    ```python
+    [42, 43]
+    ```
+    """
+
+    result = dict()
+
+    to_process: list[tuple[str, Any]] = []
+    for span_key, span_value in span._attributes.items():
+        if key == "":
+            # No key prefix, add to root
+            to_process.append((f"{key}.{span_key}", span_value))
+        elif span_key.startswith(key):
+            # Remove the key prefix and the first dot
+            to_process.append((span_key, span_value))
+
+    if not to_process:
+        if key == "":
+            # Empty span attributes
+            return result
+        raise KeyError(f"Key {key} not found in span attributes")
+
+    for span_key, span_value in to_process:
+        parts = span_key.split(".")
+        len_parts = len(parts)
+        sub_result = result
+        for idx, part in enumerate(parts):
+            if idx == len_parts - 1:
+                sub_result[part] = span_value
+            else:
+                if part not in sub_result:
+                    sub_result[part] = dict()
+                sub_result = sub_result[part]
+
+    result = _dict_to_list(result)
+    for part in key.split("."):
+        result = result[part]
+    return result
+
+
+def _linear_object(obj: dict, current: Union[list, dict, Any], key: str = ""):
```

Comment:
Could you just explain a bit why this is needed?

Reply:
I have refactored and expanded that section